common/cpp/src/google_smart_card_common/global_context_impl_emscripten.h

```diff
@@ -0,0 +1,63 @@
+// Copyright 2020 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_SMART_CARD_COMMON_GLOBAL_CONTEXT_IMPL_EMSCRIPTEN_H_
+#define GOOGLE_SMART_CARD_COMMON_GLOBAL_CONTEXT_IMPL_EMSCRIPTEN_H_
+
+#ifndef __EMSCRIPTEN__
+#error "This file should only be used in Emscripten builds"
+#endif  // __EMSCRIPTEN__
+
+#include <mutex>
+#include <thread>
+
+#include <emscripten/val.h>
+
+#include <google_smart_card_common/global_context.h>
+#include <google_smart_card_common/value.h>
+
+namespace google_smart_card {
+
+// Implementation of the GlobalContext interface for the Emscripten
+// (WebAssembly) environment.
+class GlobalContextImplEmscripten final : public GlobalContext {
+ public:
+  // `post_message_callback` - JavaScript callback that will be called for
+  // posting a message.
+  GlobalContextImplEmscripten(std::thread::id main_thread_id,
+                              emscripten::val post_message_callback);
+  GlobalContextImplEmscripten(const GlobalContextImplEmscripten&) = delete;
+  GlobalContextImplEmscripten& operator=(const GlobalContextImplEmscripten&) =
+      delete;
+  ~GlobalContextImplEmscripten() override;
+
+  // Disables posting new messages to JS. All calls to `PostMessageToJs()` after
+  // this point will return `false`.
+  void DetachFromPostMessage();
```

Comment:
BTW, maybe I should call it and the similar method in the other class simply `ShutDown()`. Happy to rename if you agree it's better.

Reply:
IMHO, maybe not `ShutDown()` as it indicates the pointer (or generally the service) will not be valid for subsequent method calls, like `PostMessageToJs()` in this case.

I have a question, though: why don't you define this common method in the interface `GlobalContext`? I see they share the same intention but differ in implementation. For the name, it can be `DetachFromPostMessage()` or `DisablePostMessageToJs()`.

Reply:
Yup, good observation regarding making this method virtual. I was originally thinking that these methods in two classes are conceptually different (one is when the NaCl framework tells us to destroy the `Instance`, and another will be used when the JS code destroys the proxy object), but, looking at them now, they're essentially about the same operation.

What about a name like `DisableJsCommunication()`? This will be a bit more future-proof, in case we add more stuff that needs to be "shut down" together with this JS-to-C++ pipe.

Reply:
yes, I like the name.

Reply:
Done.